### PR TITLE
Add policy listener support

### PR DESCRIPTION
**Pull request by jrajahalme**
Add support for `listener` keyword in CNP to redirect matching traffic to an Envoy listener.

Please review commit-by-commit.

```release-note
Traffic can now we redirected to Envoy listeners via Cilium Network Policy `listener` option.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
Rebased.

**Comment by jrajahalme** [comment]
Now using tip of `cilium/proxy` (1st commit updated).

**Comment by jrajahalme** [comment]
Dropped commit from https://github.com/cilium/cilium/pull/21805 to get past base image lint.

**Comment by jrajahalme** [comment]
I have fixed build-commits breakage locally by adding `GetListener()` definitions to the 9th (2nd to last) commit locally. Will wait for more Ci builds to complete before pushing.

File: pkg/envoy/ciliumenvoyconfig.go
Comment (chancez) [inline pkg/envoy/ciliumenvoyconfig.go:69]: Could probably use a unit test for this logic. Also, if we have a string-set data structure somewhere already, we could use it with listener names and do the set difference/intersection instead of manually calculating the difference with slices.

File: pkg/envoy/ciliumenvoyconfig_test.go
Comment (chancez) [inline pkg/envoy/ciliumenvoyconfig_test.go:729]: Is there a reason we can't just construct a full `Resource` and do a deep comparison instead of a dozen asserts on individual fields?

**Comment by chancez** [comment]
When resources can reference other resources, one thing that's often missed in this is namespacing/RBAC concerns where a resource in one namespace can be referenced by a resource in a different namespace.

I can't quite tell, but given the listener field doesn't reference a CEC or a namespace in anyway, I think this is a potential gap in the implementation. This PR should account for the case where:

- Namespace A defines CEC with a custom listener
- Namespace B defines a CiliumNetworkPolicy which sends traffic to the CEC listener defined in namespace A.

IMO, namespace B should not be able to reference listeners defined in namespace A. Currently I don't think there's anything here preventing that. 

Additionally, I wonder if you have potential issues where 2 listeners defined have conflicting names and how that can be handled. Without proper namespacing of the names, a user could "take over" a listener that someone else created in a different CEC, by trying to repeatedly create a CEC using a listener with the same name as another. The hope being the original CEC with a listener is deleted briefly, leaving enough time to create their own CEC with their listener. This is similar to port 22/443/80 being privileged ports, to avoid non-root users on a server from being able to take over port 22 with a malicious SSH server.

To address the first issue, you could change listener names so they contain the CEC and the CEC namespace in them when rendered to an envoy config. Then it would be required that the user specifies the CEC name in the policy rule, to avoid ambiguity and potential "listener takeover" I mentioned earlier. The namespace will be the same as the policy's namespace, or the `listener.envoyConfig.namespace` if using a `CiliumClusterNetworkPolicy`.

For example:
```
toPorts:
  listener:
    name: my-custom-listener
    envoyConfig:
      name: foo-cec # must specify which CEC where listener is defined
      namespace: some-ns # only required when using cluster-wide CNP

```

**Comment by chancez** [comment]
On a similar note to above, it might be wise to change `listener` to a object with a `name` field anyways, just to future proof it in case you need to add additional configuration options to the listener reference in the policy rules.

File: pkg/envoy/ciliumenvoyconfig.go
Comment (jrajahalme) [reply pkg/envoy/ciliumenvoyconfig.go:69]: Will add a unit test, but generally I'd like to keep predicates like this not allocation memory when not needed. Not that this is on a particularly hot path, but still...

File: pkg/envoy/ciliumenvoyconfig_test.go
Comment (jrajahalme) [reply pkg/envoy/ciliumenvoyconfig_test.go:729]: No reason, it was just easier.

**Comment by jrajahalme** [comment]
> On a similar note to above, it might be wise to change `listener` to a object with a `name` field anyways, just to future proof it in case you need to add additional configuration options to the listener reference in the policy rules.

I'll do this, and add a namespace field for CCNP use. Not sure how to refer to the CEC yet.. right now a `listener` must contain the namespace and CEC name in the reference, e.g. `some-ns/foo-cec/my-custom-listener`, but there is no restriction on the namespace, as you noticed.

**Comment by jrajahalme** [comment]
@chancez Added the unit testing you asked for, and changed the listener name in policy to a structured type. Added `GetNamespace()` method to `PolicyContext` to be able to default to the enclosing rule's namespace if namespace is left out. Just noticed that I did not add defaulting to "default" if rule's namespace is empty (i.e., as in CCNP). Left the this change into a separate commit so that it will be easier for you to review, will squash before merge :-)

File: pkg/policy/api/rule_validation.go
Comment (chancez) [inline pkg/policy/api/rule_validation.go:385]: We could validate that if `listener.EnvoyConfig.Namespace` is empty for CNPs (should only be specified for CCNP).

File: pkg/policy/l4.go
Comment (chancez) [inline pkg/policy/l4.go]: Should probably check listener namespace matches policy namespace (for non-CCNP's)

**Review by chancez — CHANGES_REQUESTED** [review]
I only reviewed recent commits so I may be missing something, but it seems we're not validating the Policies `listener.envoyConfig.namespace` matches the CNP namespace. 

Also, I'm a bit confused about the default of `"default"`. 

To clarify:
- Listener namespace should always be empty for CNPs referencing CECs, and the CNPs will reference CEC listeners in the same namespace.
- For CCNPs, the listener namespace should be required, and must match the namespace/name of an existing CEC.

I also just realized, we probably want to support CNPs referencing a CCEC. For that, we will need to add a `kind` field to the `listener.envoyConfig.kind` so we can disambiguate, similar to RoleBindings which can reference a Role or ClusterRole.

**Comment by jrajahalme** [comment]
@chancez Yes, CNPs using a shared CCEC defined listener is a valid use case. We currently disambiguate between CEC and CCEC defined listeners (in listener names) by the fact that CCEC do not have a namespace and CEC's do. Given this we need:

- CNPs should only be able to refer to listeners defined in CECs in their own namespace, or in CCECs for shared listeners
- CCNPs should only be able to refer to listeners defined in CCECs

For the above a namespace in a CNP or CCNP Listener reference is not useful at all.

I'm also a bit ambivalent about using the `envoyConfig` keyword in the CNP/CCNP. IMO something more generic would be better. Thoughts/opinions/ideas?

**Comment by jrajahalme** [comment]
> I'm also a bit ambivalent about using the `envoyConfig` keyword in the CNP/CCNP. IMO something more generic would be better. Thoughts/opinions/ideas?

Or then make it even more explicit, and in addition to removing the `namespace` reference (that is not needed), replace `envoyConfig` with these two, out of which one must be specified:

- cec: Name of a CiliumEnvoyConfig resource in the same namespace as a CNP (only valid for a CNP)
- ccec: Name of a CiliumClusterwideEnvoyConfig resource

Alternative names for these two could be the current `envoyConfig` and `sharedEnvoyConfig`, or more abstractly, `resource` and `clusterResource`, or some variation thereof.

**Comment by jrajahalme** [comment]
@chancez Added a new commit addressing the namespacing issue you pointed out. This removes `namespace` from the listener reference, as the listener must either be in the same namespace as the rule (for CNP rules), or be non-namespaced listener (always for CCNP, optionally for CNP rules). To achieve this I added a `clusterwideEnvoyConfig` that can be used in place of `envoyConfig`, when referring to a listener defined in (non-namespaced) ClusterwideCiliumEnvoyConfig.

I still hope we can come up with better names for `envoyConfig` and `clusterwideEnvoyConfig` fields, if possible.

**Comment by chancez** [comment]
@jrajahalme yeah this looks great. I'm not sure what's best for naming, I think the main things we should consider is if we will need to extend this in the future, then a nested structure is better, as it will let us add more fields easily later.

I also think we could still use a single `envoyConfig` field rather than a `sharedEnvoyConfig`, and just add `kind` field to it, like RoleBindings do with roleReferences. Eg:
```
```

**Comment by jrajahalme** [comment]
> I still think we could still use a single envoyConfig field rather than a clusterwideEnvoyConfig, and just add kind field to it, like RoleBindings do with roleReferences. Eg:
> ```listener:
>   envoyConfig:
>     name: foo-cec
>     kind: CiliumClusterwideEnvoyConfig # optional, if unset, it's equal to kind: CiliumEnvoyConfig

@chancez It would seem `kind` should default to `CiliumEnvoyConfig` for CNPs and `CiliumClusterwideEnvoyConfig` for CCNPs, especially since reference to a CEC is not allowed from a CCNP?

**Comment by jrajahalme** [comment]
@chancez Added explicit `kind` to `envoyConfig`, squashed late commits together to be ready for merging.

**Review by chancez — APPROVED** [review]
LGTM

**Review by aanm — APPROVED** [review]

File: pkg/endpoint/bpf.go
Comment (christarazi) [inline pkg/endpoint/bpf.go]: Previously this was an error path, but now it's no longer returning errors here. Could you elaborate on the change of behavior here and what the consequences are? Is debug level enough for logging?

File: pkg/k8s/watchers/cilium_clusterwide_envoy_config.go
Comment (christarazi) [inline pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:121]: Do you intend to resolve this in this PR? Otherwise, we should create a GH issue for follow ups and put this item there, while referencing the GH issue in the code comment.

Comment (christarazi) [reply pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:121]: Also, I'm not following "Policy does not need to be recomputed for this, but if we do not 'force' the bpf maps are not updated with the new proxy ports either". AFAIU, policy needs to be recomputed in order for there to be a change to update in the BPF maps, so this `TriggerPolicyUpdates` sounds like it is needed in contrast to what the comment says :slightly_smiling_face:.

File: pkg/policy/api/l4.go
Comment (christarazi) [inline pkg/policy/api/l4.go:147]: If you don't expect this to be empty, then we can add validation on length so that it's rejected if empty.

[1]

File: pkg/policy/api/rule_validation.go
Comment (christarazi) [inline pkg/policy/api/rule_validation.go]: [1] will allow you to drop this line.

File: pkg/policy/api/rule_validation.go
Comment (christarazi) [inline pkg/policy/api/rule_validation.go]: This error msg suggests that empty is not a valid `kind`, but the case above handles it. Also, the godoc on the `Kind` type above also suggests that it's optional (which means the field can be empty). So which is it? :slightly_smiling_face:

File: pkg/policy/api/rule_validation.go
Comment (christarazi) [inline pkg/policy/api/rule_validation.go]: Couple of TODOs here. Do you intend to implement in this PR? Apply follow up comment from above if not.

File: pkg/policy/l4.go
Comment (christarazi) [inline pkg/policy/l4.go:744]: Unfortunately, this may not be possible to do at CRD-level validation, because the CCNP type is embedded inside the CNP type. This means that the validation is simply copied over to CCNP. This has it's pros and cons, because 99% of cases, a CCNP should be validated like a CNP as they're pretty much the same thing, except when they are not. 

It's also not as easy as just not embedding the type because there are a lot of places in the code which assumes this structure, last I looked (>1y).

**Review by christarazi — COMMENTED** [review]
Great work! A few comments below.

File: pkg/endpoint/bpf.go
Comment (jrajahalme) [reply pkg/endpoint/bpf.go]: This is expected to happen when policy refers to a listener that does not exist yet, for example when restarting and k8s delivers the CNP before the related CEC. Related to this policy regeneration is now triggered when listeners are added/changed/deleted.

Debug level is enough to see that this happens as intended.

File: pkg/k8s/watchers/cilium_clusterwide_envoy_config.go
Comment (jrajahalme) [reply pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:121]: Not planning for this PR, as this related also to other cases where policy regeneration is forced. In this particular case we'd only need the existing endpoint policy be synced to bpf policy maps, as that is where the procy redirection port is resolved. This is why the endpoint policy itself would not need to be recomputed.

File: pkg/policy/api/l4.go
Comment (jrajahalme) [reply pkg/policy/api/l4.go:147]: Missed this, thought that would be part of required... ofcause it is not!

File: pkg/policy/api/rule_validation.go
Comment (jrajahalme) [reply pkg/policy/api/rule_validation.go]: Empty is allowed, with sane defaulting as I think I wrote somewhere.

File: pkg/policy/l4.go
Comment (jrajahalme) [reply pkg/policy/l4.go:744]: I meant `pkg/policy/api/rule_validation.go` with this, not k8s client-side validation. On our own policy API this would need passing in context w.r.t. what kind of policy in being validated. That is somewhat large refactor, so did not attempt it in this PR.

File: pkg/endpoint/bpf.go
Comment (christarazi) [reply pkg/endpoint/bpf.go]: I think 

> for example when restarting and k8s delivers the CNP before the related CEC

is good to add to the code comment, because otherwise it's not clear why.

File: pkg/k8s/watchers/cilium_clusterwide_envoy_config.go
Comment (christarazi) [reply pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:121]: > In this particular case we'd only need the existing endpoint policy be synced to bpf policy maps, as that is where the procy redirection port is resolved.

So there are no BPF policymap updates needed here, but the proxy redirection port logic is embedded somewhere in the policy regeneration / recalculation path (therefore that's why a full policy recalculation works)? And we should only be triggering proxy redirection port logic and not a full regeneration / recalculation? How difficult is it to pull that out, just out of curiosity?

File: pkg/endpoint/bpf.go
Comment (jrajahalme) [reply pkg/endpoint/bpf.go]: Reworded as:
```
	// Skip redirects that can not be created or updated.  This
	// can happen when a listener is missing, for example when
	// restarting and k8s delivers the CNP before the related
	// CEC.
	// Policy is regenerated when listeners are added or removed
	// to fix this condition when the listener is available.
	e.getLogger().WithField(logfields.Listener, l4.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
	continue
```

File: pkg/k8s/watchers/cilium_clusterwide_envoy_config.go
Comment (jrajahalme) [reply pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:121]: BPF policy maps need to be updated so that the previously skipped redirect rules are pushed to the bpf maps (see your first comment). We would need to push more context info with the regeneration trigger to be able to skip parts of the regeneration that are known to not be needed. We already do something similar to the BPF code generation, where we can skip compilation and only load the new programs, etc.

File: pkg/policy/api/l4.go
Comment (jrajahalme) [reply pkg/policy/api/l4.go:147]: Added:
```
	// +kubebuilder:validation:MinLength=1
```

File: pkg/policy/api/rule_validation.go
Comment (jrajahalme) [reply pkg/policy/api/rule_validation.go]: Should I drop all other (manual) validation in favor of letting k8s worry about it as well?

Comment (jrajahalme) [reply pkg/policy/api/rule_validation.go]: Removed also other validation that is already done vie kubebuilder validation

File: pkg/policy/api/rule_validation.go
Comment (jrajahalme) [reply pkg/policy/api/rule_validation.go]: Added wording for "follow-up work" for both of these.

**Comment by jrajahalme** [comment]
@christarazi Addressed your comments and rebased :-)

**Review by christarazi — APPROVED** [review]

**Pull request merged**
